File: bindings/typescript/tests/persistence.test.ts
/**
 * SDK-layer smoke test for SQLite-backed persistence (#1549 Phase 4 PR 3).
 *
 * Verifies that the TypeScript SDK wrapper surface:
 *
 * 1. Accepts the documented `StorageConfig.sqlite` variant and forwards it
 *    to the native NAPI `SCP.withStorage(configJson)` factory without
 *    raising.
 * 2. Creates the SQLCipher database file at `{path}/scp.db` as a side
 *    effect of construction — that is what the FFI
 *    `StorageConfig::Sqlite` path does (see
 *    `crates/scp-ffi/napi/src/runtime.rs::with_storage_napi`).
 * 3. Drives the full `suspend() → resume() → shutdown()` lifecycle on a
 *    SQLite-backed instance without error.
 * 4. Is reconstructible against the SAME SQLite path + key — the
 *    reopened instance must be able to open the encrypted database
 *    again without re-deriving a fresh key.
 *
 * The wrapper surface is all this smoke test is responsible for. The
 * end-to-end `identity_create → context_create → context_send → suspend
 * → restore` path is exercised at the Rust integration layer
 * (`crates/scp-testing/tests/integration/persistence_sdk.rs`) because
 * the TypeScript `SCP` class does not yet surface context methods.
 * Phase 4 PR 4 (#1549, ADR-048) deleted the free-function façade
 * (`contextCreate`, etc.) along with the process-wide default bridge
 * it routed through; the remaining SDK wiring to expose those
 * operations on the caller-owned `SCP` handle is tracked as follow-up
 * work.
 *
 * Skipped at file level when the native NAPI addon is unavailable
 * (browser runtime, missing platform binary).
 */

import { describe, expect, it } from "bun:test";
import { existsSync } from "node:fs";
import { mkdtemp, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { SCP } from "../src/scp";

// Stable 32-byte SQLCipher key. The specific value does not matter;
// only that the same key is reused across the two SCP constructions
// that simulate process restart.
const SQLITE_KEY = new Uint8Array(32).fill(0x42);

async function makeTempDir(): Promise<string> {
  return await mkdtemp(join(tmpdir(), "scp-persistence-"));
}

async function withTempDir<T>(fn: (dir: string) => Promise<T>): Promise<T> {
  const dir = await makeTempDir();
  try {
    return await fn(dir);
  } finally {
    await rm(dir, { recursive: true, force: true });
  }
}

// Best-effort detection of whether the NAPI addon is available. We
// attempt a cheap `new SCP({ storage: { type: "in_memory" } })` inside a try/catch — if the addon is
// structurally unavailable we'll get a `SCP-VALID-7005` ValidationError
// and skip the whole suite. This keeps the test file runnable in
// environments without the native addon (e.g. an unsupported platform)
// without hard-failing.
function napiAvailable(): boolean {
  try {
    const probe = new SCP({ storage: { type: "in_memory" } });
    probe.shutdown(1).catch(() => {});
    return true;
  } catch {
    return false;
  }
}

const describeNapi = napiAvailable() ? describe : describe.skip;

describeNapi("SCP with SQLite storage (#1549 PR 3)", () => {
  it("creates the SQLCipher database file at {path}/scp.db", async () => {
    await withTempDir(async (dir) => {
      const dbPath = join(dir, "scp.db");
      expect(existsSync(dbPath)).toBe(false);

      const scp = new SCP({
        storage: { type: "sqlite", path: dir, key: SQLITE_KEY },
      });
      try {
        expect(existsSync(dbPath)).toBe(true);
        expect(scp.instanceId).toBeDefined();
      } finally {
        await scp.shutdown(1);
      }
    });
  });

  it("roundtrips suspend / async resume / shutdown on a SQLite-backed instance", async () => {
    await withTempDir(async (dir) => {
      const scp = new SCP({
        storage: { type: "sqlite", path: dir, key: SQLITE_KEY },
      });
      try {
        scp.suspend();
        await scp.resume();
      } finally {
        await scp.shutdown(1);
      }
    });
  });

  it("reopens the same path+key across two SCP constructions", async () => {
    await withTempDir(async (dir) => {
      const config = { type: "sqlite" as const, path: dir, key: SQLITE_KEY };

      const scp1 = new SCP({ storage: config });
      const id1 = scp1.instanceId;
      await scp1.shutdown(1);

      const scp2 = new SCP({ storage: config });
      try {
        // Compare as BigInts to sidestep JS's 53-bit mantissa when the
        // native instance_id is exposed as a string or BigInt. The
        // only property we actually care about is "the construction
        // succeeded" — strict monotonicity is covered elsewhere.
        expect(scp2.instanceId).toBeDefined();
        expect(scp2.instanceId).not.toEqual(id1);
      } finally {
        await scp2.shutdown(1);
      }
    });
  });

  it("fails closed on a mismatched key without corrupting the original DB", async () => {
    await withTempDir(async (dir) => {
      // First open with the correct key — creates the encrypted DB.
      const scp1 = new SCP({
        storage: { type: "sqlite", path: dir, key: SQLITE_KEY },
      });
      await scp1.shutdown(1);

      // Second open with a WRONG key. FAIL CLOSED (spec §17.6): the NAPI
      // bridge surfaces the SQLCipher key-mismatch as a `SCP-STORAGE-8001`
      // validation error rather than silently returning an in-memory
      // instance. All three bridges report that one code for a failed
      // durable-backend open, so a caller reads the same code whichever
      // binding raised it. The original encrypted DB must survive the failed
      // attempt so the next correct-key open still works.
      const wrongKey = new Uint8Array(32).fill(0x11);
      expect(() => {
        new SCP({
          storage: { type: "sqlite", path: dir, key: wrongKey },
        });
      }).toThrow(/SCP-STORAGE-8001/);

      // Third open with the correct key — must still succeed, proving
      // the failed mismatched-key attempt did not corrupt or truncate
      // the encrypted database file.
      const scp3 = new SCP({
        storage: { type: "sqlite", path: dir, key: SQLITE_KEY },
      });
      await scp3.shutdown(1);
    });
  });

  it("round-trips a passphrase-keyed SQLite instance across two constructions", async () => {
    await withTempDir(async (dir) => {
      const passphrase = "correct horse battery staple";

      // First open with a passphrase — creates the encrypted DB and the
      // persisted Argon2id salt sidecar.
      const scp1 = new SCP({
        storage: { type: "sqlite", path: dir, passphrase },
      });
      expect(scp1.instanceId).toBeDefined();
      await scp1.shutdown(1);

      // Reopen with the SAME passphrase — must succeed (the salt sidecar
      // re-derives the same SQLCipher key).
      const scp2 = new SCP({
        storage: { type: "sqlite", path: dir, passphrase },
      });
      try {
        expect(scp2.instanceId).toBeDefined();
      } finally {
        await scp2.shutdown(1);
      }
    });
  });

  it("fails closed when reopening a passphrase DB with the wrong passphrase", async () => {
    await withTempDir(async (dir) => {
      // Create with the correct passphrase.
      const scp1 = new SCP({
        storage: { type: "sqlite", path: dir, passphrase: "the-right-passphrase" },
      });
      await scp1.shutdown(1);

      // Reopen with the WRONG passphrase. FAIL CLOSED (spec §17.6):
      // SQLCipher rejects the derived key — the NAPI layer must throw,
      // NOT silently open a fresh, empty database.
      expect(() => {
        new SCP({
          storage: { type: "sqlite", path: dir, passphrase: "the-WRONG-passphrase" },
        });
      }).toThrow();
    });
  });

  it("rejects sqlite config that supplies both key and passphrase", () => {
    // The NAPI layer enforces the mutual exclusion at the JSON boundary
    // (SCP-VALID-7005). We bypass the TS union type (which models the two
    // sqlite shapes separately) with a cast to assert the runtime guard.
    expect(() => {
      new SCP({
        storage: {
          type: "sqlite",
          path: "/tmp/scp-both",
          key: SQLITE_KEY,
          passphrase: "also-a-passphrase",
        } as unknown as { type: "sqlite"; path: string; passphrase: string },
      });
    }).toThrow();
  });
});

File: bindings/kotlin/scp-kt/src/test/kotlin/works/limn/scp/PersistenceTest.kt
// PersistenceTest.kt — SDK-layer smoke test for SQLite-backed persistence.
//
// #1549 Phase 4 PR 3 + PR 4. Verifies that the Kotlin SDK wrapper surface:
//
// 1. Accepts `SCP.withSqlite(dir, key)` and forwards to UniFFI
//    `Scp.withStorage(StorageConfig.Sqlite(path, key))` without raising.
// 2. Creates the SQLCipher database file at `{dir}/scp.db` as a side
//    effect of construction — see
//    `crates/scp-ffi/uniffi/src/runtime.rs::with_storage_uniffi`.
// 3. Drives the full `suspendInstance() → resume() → shutdown()` lifecycle
//    on a SQLite-backed instance without error. All three are `suspend`
//    functions in the Kotlin SDK, routed through `CoroutineBridge`.
// 4. Is reconstructible against the SAME SQLite directory + key — the
//    reopened instance must open the encrypted database again without
//    re-deriving a fresh key.
//
// Each test gets its own tmp dir + `SCP.withSqlite(...)` instance and
// tears it down deterministically via `@AfterEach` so tests don't leak
// runtime state across the suite.
//
// The end-to-end `identityCreate → contextCreate → contextSend → suspend
// → restore` path is exercised at the Rust integration layer
// (`crates/scp-testing/tests/integration/persistence_sdk.rs`).
//
// The suite skips (via JUnit 5 `assumeTrue`) when the UniFFI cdylib
// is not loadable, matching the pattern in `ScpClassTest`.

package works.limn.scp

import java.io.File
import java.nio.file.Files
import kotlin.io.path.exists
import kotlin.io.path.pathString
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue
import uniffi.scp.ScpException
import kotlin.time.Duration.Companion.seconds
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assumptions.assumeTrue
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import works.limn.scp.bridge.CoroutineBridge
import works.limn.scp.conformance.ConformanceStubBindings

@OptIn(ExperimentalCoroutinesApi::class)
class PersistenceTest {
    companion object {
        /**
         * Error code the UniFFI bridge raises when a durable storage backend
         * cannot be opened — `StorageInitError::SqliteOpen` converted to
         * `ScpError::Validation` in `crates/scp-ffi/uniffi/src/runtime.rs`, the
         * same code the `PyO3` and NAPI bridges report for that same failure.
         * Spec §17.6 makes a failed durable-backend open terminal, so this code
         * is the observable proof that no in-memory downgrade happened.
         */
        private const val STORAGE_OPEN_FAILED_CODE = "SCP-STORAGE-8001"

        private var nativeAvailable = false
        private var skipReason = ""

        @JvmStatic
        @BeforeAll
        fun probeNativeLibrary() {
            try {
                Class.forName("uniffi.scp.ScpKt")
                Class.forName("uniffi.scp.Scp\$Companion")
                nativeAvailable = true
            } catch (e: ClassNotFoundException) {
                skipReason = "UniFFI bindings not available: ${e.message}"
            } catch (e: UnsatisfiedLinkError) {
                skipReason = "Native library link error: ${e.message}"
            } catch (e: ExceptionInInitializerError) {
                skipReason = "Native library init error: ${e.cause?.message ?: e.message}"
            } catch (e: NoClassDefFoundError) {
                skipReason = "Native library class not found: ${e.message}"
            }
        }
    }

    /// Stable 32-byte SQLCipher key. The specific value does not matter;
    /// only that the same key is reused across the two constructions
    /// that simulate process restart.
    private val sqliteKey: ByteArray = ByteArray(32) { 0x42 }

    private val createdInstances = mutableListOf<SCP>()

    private fun bridge(): CoroutineBridge =
        CoroutineBridge(
            nativeBindings = ConformanceStubBindings(),
            ioDispatcher = Dispatchers.IO,
            cpuDispatcher = Dispatchers.Default,
        )

    @BeforeEach
    fun requireNative() {
        assumeTrue(nativeAvailable, skipReason)
    }

    @AfterEach
    fun tearDown() {
        val b = bridge()
        for (scp in createdInstances) {
            runBlocking { scp.shutdown(b, 1.seconds) }
        }
        createdInstances.clear()
    }

    private fun makeSqliteScp(dir: File): SCP {
        val scp = SCP.withSqlite(dir, sqliteKey)
        createdInstances += scp
        return scp
    }

    @Test
    fun `withSqlite creates the SQLCipher database file at dir slash scp dot db`() =
        runTest {
            val dir = Files.createTempDirectory("scp-kotlin-persist-")
            dir.toFile().deleteOnExit()
            val dbPath = dir.resolve("scp.db")
            assertTrue(
                !dbPath.exists(),
                "scp.db must not exist before SCP.withSqlite at ${dbPath.pathString}",
            )

            val scp = makeSqliteScp(dir.toFile())

            assertTrue(
                dbPath.exists(),
                "SCP.withSqlite(dir, key) must create scp.db at ${dbPath.pathString}",
            )
            assertNotEquals(0UL, scp.instanceId, "instanceId must be monotonic and non-UNSET")
        }

    @Test
    fun `withSqlite roundtrips suspendInstance then resume via CoroutineBridge`() =
        runTest {
            val dir = Files.createTempDirectory("scp-kotlin-persist-lifecycle-")
            dir.toFile().deleteOnExit()
            val scp = makeSqliteScp(dir.toFile())

            scp.suspendInstance(bridge())
            scp.resume(bridge())
            // Reaching here means both suspend and resume completed
            // without raising — the SQLite-backed path composes with
            // the async resume wiring introduced in #1549 PR 3.
            assertTrue(true)
        }

    @Test
    fun `withSqlite reopens the same dir plus key across two constructions`() =
        runTest {
            val dir = Files.createTempDirectory("scp-kotlin-persist-reopen-")
            dir.toFile().deleteOnExit()

            val scp1 = makeSqliteScp(dir.toFile())
            val id1 = scp1.instanceId
            // Post-9fa80e13c: SqliteStorage::new propagates open failures
            // instead of falling back to in-memory, so the first instance
            // must release the DB before a second open can succeed. This
            // matches the Swift `testSqliteReopenWithSamePathAndKeySucceeds`
            // shape.
            scp1.shutdown(bridge(), 1.seconds)
            createdInstances.remove(scp1)

            val scp2 = makeSqliteScp(dir.toFile())
            val id2 = scp2.instanceId

            assertNotEquals(
                id1,
                id2,
                "each SCP.withSqlite(...) must produce an independent UniffiBridgeInstance",
            )
        }

    @Test
    fun `withSqlite fails closed on mismatched key without corrupting the DB`() =
        runTest {
            val dir = Files.createTempDirectory("scp-kotlin-persist-mismatch-")
            dir.toFile().deleteOnExit()

            // First open with the correct key — creates the encrypted DB.
            val scp1 = makeSqliteScp(dir.toFile())
            scp1.shutdown(bridge(), 1.seconds)
            createdInstances.remove(scp1)

            // Second open with a wrong key MUST throw — `SqliteStorage::new`
            // fails at the `PRAGMA key` / WAL-mode step because `SQLCipher`
            // rejects the key as "file is not a database". The UniFFI bridge
            // converts that `StorageInitError::SqliteOpen` to
            // `ScpError::Validation` with code `SCP-STORAGE-8001`
            // (`crates/scp-ffi/uniffi/src/runtime.rs`, the
            // `From<StorageInitError> for ScpError` impl), which the generated
            // Kotlin surfaces as `ScpException.Validation`. Main's 9fa80e13c
            // replaced the former silent fallback to in-memory (split-brain
            // where writes silently vanished) with hard-error propagation.
            val wrongKey = ByteArray(32) { 0x11 }
            val rejected =
                assertFailsWith<ScpException.Validation> {
                    SCP.withSqlite(dir.toFile(), wrongKey)
                }
            assertEquals(
                STORAGE_OPEN_FAILED_CODE,
                rejected.code,
                "a rejected SQLCipher key is a storage-open failure, not a context failure",
            )

            // Third open with the correct key — must still succeed, proving
            // the failed mismatched-key attempt did not corrupt or truncate
            // the encrypted database file.
            makeSqliteScp(dir.toFile())
        }

    @Test
    fun `withSqlite passphrase round trips across two constructions`() =
        runTest {
            val dir = Files.createTempDirectory("scp-kotlin-persist-passphrase-")
            dir.toFile().deleteOnExit()
            val dbPath = dir.resolve("scp.db")

            val scp1 = SCP.withSqlite(dir.toFile(), passphrase = "correct horse battery staple")
            assertTrue(
                dbPath.exists(),
                "passphrase construction must create scp.db at ${dbPath.pathString}",
            )

            // `SqliteStorage` holds a process-exclusive advisory lock on
            // `scp.db.lock` for its lifetime, so close the first handle before
            // the second construction. Leaving it open tested lock exclusion,
            // not the passphrase round trip this test is named for. The sibling
            // test `withSqlite reopens the same dir plus key across two
            // constructions` shuts down for the same reason.
            scp1.shutdown(bridge(), 1.seconds)

            // Reopen with the SAME passphrase — must succeed (salt sidecar
            // re-derives the same key).
            val scp2 = SCP.withSqlite(dir.toFile(), passphrase = "correct horse battery staple")
            createdInstances += scp2
        }

    @Test
    fun `withSqlite fails closed on wrong passphrase`() =
        runTest {
            val dir = Files.createTempDirectory("scp-kotlin-persist-wrongpass-")
            dir.toFile().deleteOnExit()

            val scp1 = SCP.withSqlite(dir.toFile(), passphrase = "the-right-one")
            // Release the advisory lock on `scp.db.lock` first, so the reopen
            // below fails on the WRONG PASSPHRASE and not on a still-held lock.
            // While the first handle stayed open, that lock rejected the second
            // construction, so the wrong passphrase decided nothing and this
            // test proved only that two handles cannot coexist.
            scp1.shutdown(bridge(), 1.seconds)

            // Reopen with the WRONG passphrase must fail closed — never
            // silently open a fresh DB (spec §17.6).
            val rejected =
                assertFailsWith<ScpException.Validation> {
                    SCP.withSqlite(dir.toFile(), passphrase = "the-WRONG-one")
                }
            assertEquals(
                STORAGE_OPEN_FAILED_CODE,
                rejected.code,
                "a rejected passphrase is a storage-open failure, not a context failure",
            )
        }
}
